ssh-keygen.c

```diff
@@ -372,20 +376,44 @@ static void
 do_convert_to_pkcs8(struct sshkey *k)
 {
 	switch (sshkey_type_plain(k->type)) {
+#if (OPENSSL_VERSION_NUMBER >= 0x30000000L)
 	case KEY_RSA:
-		if (!PEM_write_RSA_PUBKEY(stdout, k->rsa))
-			fatal("PEM_write_RSA_PUBKEY failed");
+	case KEY_ECDSA:
+		OSSL_ENCODER_CTX *ctx = OSSL_ENCODER_CTX_new_for_pkey(k->pkey,
+			                                              EVP_PKEY_PUBLIC_KEY,
+			                                              "PEM",
+                                                          	      "type-specific",
+                                                          	      NULL);
+		if (!ctx)
+			fatal("OSSL_ENCODER_CTX_new_for_pkey failed");
+		if (OSSL_ENCODER_to_fp(ctx, stdout) != 1) {
+			OSSL_ENCODER_CTX_free(ctx);
+			fatal("OSSL_ENCODER_to_fp failed");
+		}
+		OSSL_ENCODER_CTX_free(ctx);
 		break;
-	case KEY_DSA:
-		if (!PEM_write_DSA_PUBKEY(stdout, k->dsa))
-			fatal("PEM_write_DSA_PUBKEY failed");
+#else
+	case KEY_RSA:
+		{
+			RSA *rsa = EVP_PKEY_get0_RSA(k->pkey);
+			if (!PEM_write_RSA_PUBKEY(stdout, rsa))
+				fatal("PEM_write_RSA_PUBKEY failed");
+		}
 		break;
 #ifdef OPENSSL_HAS_ECC
 	case KEY_ECDSA:
-		if (!PEM_write_EC_PUBKEY(stdout, k->ecdsa))
-			fatal("PEM_write_EC_PUBKEY failed");
+		{
+			EC_KEY *ec = EVP_PKEY_get0_EC_KEY(k->pkey);
+			if (!PEM_write_EC_PUBKEY(stdout, ec))
+				fatal("PEM_write_EC_PUBKEY failed");
+		}
 		break;
 #endif
+#endif
+	case KEY_DSA:
+		if (!PEM_write_DSA_PUBKEY(stdout, k->dsa))
+			fatal("PEM_write_DSA_PUBKEY failed");
+		break;
 	default:
 		fatal_f("unsupported key type %s", sshkey_type(k));
 	}
```

Comment:
I would replace all of this with:

``` C
	if (k->pkey != NULL) {
		if (!PEM_write_PUBKEY(stdout, k->pkey))
			fatal("PEM_write_PUBKEY failed");
	} else {
		fatal_f("unsupported key type %s", sshkey_type(k));
	}
```

(that said, without knowing the proper conditions for "unsupported key", please adjust appropriately)

Reply:
Side note: `do_convert_to_pkcs8` is a surprising name for a function that outputs SubjectPublicKeyInfo in PEM form.  There's no PKCS#8 in sight...

Reply:
@levitte many thanks! It made the code much more compact.